### PR TITLE
Add `Costar`, more `Star` instances

### PR DESCRIPTION
**Pull request by garyb**

### CODE REVIEW

File: src/Data/Profunctor/Costar.purs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  imap = imapF
+
+instance applyCostar :: Apply (Costar f a) where
+  apply (Costar f) (Costar g) = Costar \a -> f a (g a)
+
+instance applicativeCostar :: Applicative (Costar f a) where
+  pure a = Costar \_ -> a
+
+instance bindCostar :: Bind f => Bind (Costar f a) where
+  bind (Costar m) f = Costar \x -> unCostar (f (m x)) x
```
Comment (paf31) [inline src/Data/Profunctor/Costar.purs:46]: Why do you need the `Bind f` constraint?

**Comment by paf31** [comment]
👍 Looks good, just see my one comment.

**Comment by paf31** [comment]
Oops, sorry, wrong button

File: src/Data/Profunctor/Costar.purs (hunk 1 above)
Comment (garyb) [inline src/Data/Profunctor/Costar.purs:46]: Ah, you do not :smile:

**Comment by garyb** [comment]
Ok, dropped the unnecessary `Bind` and `Monad` constraints for `Costar` instances.

**Comment by paf31** [comment]
`Costar` can have `Strong` and `Choice` too. The `Strong` instance should be straightforward, but there are two instances for `Choice`, the one from `Comonad`, and one from `Traversable` in Haskell's `profunctors`.

**Comment by garyb** [comment]
I wasn't sure about that - I was going to attempt implementing them, but noticed they weren't included in the haskell profunctors library.

**Comment by paf31** [comment]
I'm not sure why they're not included. It seems they satisfy all the right laws.

**Comment by garyb** [comment]
I've just seen your edit, I'll put something together.

**Comment by garyb** [comment]
Is this the kind of thing you had in mind for the `Strong` instance?

``` purescript
instance strongCostar :: Comonad f => Strong (Costar f) where
  first (Costar f) = Costar \x -> Tuple (f (map fst x)) (snd (extract x))
  second (Costar f) = Costar \x -> Tuple (fst (extract x)) (f (map snd x))
```

I feel like I'm missing something obvious, but `Comonad` seemed necessary.

**Comment by paf31** [comment]
That looks right. It should be the same as the `Arrow (Cokleisli f)` instance, I think.

**Comment by garyb** [comment]
Ah right yep, thanks. That's what we have there.

Maybe we should go with the `Comonad` version for `Choice` here, if the `Traversable` version is considered questionable? I guess it's questionable due to the lack of laws for traversable.

**Comment by paf31** [comment]
Ok sounds good.

**Comment by garyb** [comment]
Ok, I've added the Kleisli instances, and `Strong` and `Choice` for `Costar`.

I don't really understand how the `Costar` instance works, but after getting sucked into a rabbit hole for a while that's what the `ArrowChoice` instance for `Cokleisli` uses in Haskell, and I couldn't figure anything else out myself. It seems like it shouldn't work, since using `+++` this way implies the `Choice Costar` instance already exists.

Once this is sorted out, I'm done playing with the core libraries (aside from datetime, but I can do that after), so good to go with the 0.9.1 release :tada:

**Pull request merged**